test/smoketest_artman.py

```diff
@@ -16,98 +16,117 @@
-    ]
-
-    artman_config_whitelist = []
-    if apis:
-        for api in apis.split(','):
-            artman_config_whitelist.append('artman_%s.yaml' % api)
+CONFIG_BLACKLIST = [
+    'google/spanner/admin/instance/artman_spanner_admin_instance.yaml',
+    'google/spanner/admin/database/artman_spanner_admin_database.yaml',
+    'google/bigtable/admin/artman_bigtable_admin.yaml',
+    'google/genomics/artman_genomics.yaml',
+    'google/logging/artman_logging.yaml',
+    'google/iam/artman_iam_admin.yaml',
+]
 
-    for artman_config_dir in artman_config_dirs:
-        _smoke_test(input_dir, artman_config_dir, artman_config_whitelist)
+logger = logging.getLogger('smoketest')
+logger.setLevel(logging.INFO)
 
 
-def _smoke_test(input_dir, artman_config_dir, artman_config_whitelist):
+def run_smoke_test(root_dir, log):
+    log_file = _setup_logger(log)
     failure = []
-    for root, dirs, files in os.walk(artman_config_dir):
-        for f in fnmatch.filter(files, 'artman_*.yaml'):
-            if f in ARTMAN_CONFIG_BLACKLIST:
-                # Do not run generation tests for those in the blacklist.
-                continue
-            if artman_config_whitelist and f not in artman_config_whitelist:
-                # If apis list is given, only test those in the list
-                continue
-            api_name = re.search('artman_(.*)\.yaml', f).group(1)
-            filename = os.path.join(root, f)
-            content = open(filename).read()
-            for lang in SUPPORTED_LANGS:
-                artifact_name = '%s_gapic' % lang
-                if 'name: %s' % lang in content:
-                    if _generate_gapic_library(
-                          filename, artifact_name, input_dir):
-                        failure.append('Failed to generate gapic %s library '
-                                       'for %s' % (lang, api_name))
+    success = []
+    for artman_yaml_path in glob.glob('%s/google/**/artman_*.yaml' % root_dir,
+                                      recursive=True):
+        if os.path.relpath(artman_yaml_path, root_dir) in CONFIG_BLACKLIST:
+            logger.info('Skip %s' % artman_yaml_path)
+            continue
+        artman_config = _parse(artman_yaml_path)
+        for artifact in artman_config.artifacts:
+            logger.info('Start artifact generation for %s of %s'
+                        % (artifact.name, artman_yaml_path))
+            if _generate_gapic_library(artman_yaml_path,
+                                       artifact.name,
+                                       root_dir,
+                                       log_file):
+                msg = 'Failed to generate %s of %s.' % (
+                    artifact.name, artman_yaml_path)
+                failure.append(msg)
+            else:
+                msg = 'Succeded to generate %s of %s.' % (
+                    artifact.name, artman_yaml_path)
+                success.append(msg)
+            logger.info(msg)
```

Comment:
It seems like this double-logs everything? (line 73 + 75/77)

Reply:
It is intended. Because the test takes too long to complete and I usually want to get an update in the mid, I found it is helpful to get a peek before the test is fully complete. Actually, circleci will by default kill the task after 10mins if nothing prints out. While that is time threshold configurable, I find it is easier to just print it out at each iteration, and if one generation task does hold longer than 10 mins, it probably means there is something wrong, and I want the task to stop after 10 mins instead of waiting even longer.

Reply:
In that case, let's just `logger.info` or `logger.error` inline; no need to repeat them at the end.

Reply:
The part in the end is a summary. The smoketest writes everything to a log that will be viewable in circleci artifact section. That part includes both messaged logged by logger plus artman execution output (see the subprocess.call part). However, only the logger part is printed out to console/streamhandler. A full report without summary in the end is almost unreadable in this case.

Reply:
Ack. Did not fully understand what the log will look like. That seems reasonable.

Reply:
A report in console: https://circleci.com/gh/ethanbao/artman/171?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

A full report in artifacts: https://171-57229230-gh.circle-artifacts.com/0/tmp/reports/smoketest.log